fhir-cql-rest/src/main/java/com/ibm/fhir/cql/engine/rest/retrieve/RestFhirRetrieveProvider.java

```diff
@@ -0,0 +1,148 @@
+        }
+
+        List<Object> results = new ArrayList<>();
+        resources.stream().forEach(resource -> {
+            if (resource instanceof Bundle) {
+                FhirBundleCursor cursor = new FhirBundleCursor(url -> {
+                    try {
+                        Response response = fhirClient.getWebTarget().path(url).request().get();
+                        FHIRClientUtil.handleErrorResponse(response);
+                        return response.readEntity(Bundle.class);
+                    } catch (Exception ex) {
+                        throw new RuntimeException(ex);
+                    }
+                }, (Bundle) resource);
+                cursor.forEach(results::add);
+            } else {
+                results.add(resource);
+            }
+        });
+
+        return results;
+    }
+
+    protected Resource executeQuery(String dataType, SearchParameterMap map) throws Exception {
+        Resource result = null;
+
+        FHIRParameters query = getFHIRParameters(map);
+
+        if (query.getParameterMap().containsKey("_id")) {
+            FHIRResponse response = fhirClient.read(dataType, query.getParameterMap().getFirst("_id"));
+            FHIRClientUtil.handleErrorResponse(response);
+
+            Class<? extends Resource> clazz = ModelSupport.getResourceType(dataType);
+            result = response.getResource(clazz);
+        } else {
+            FHIRResponse response = fhirClient.search(dataType, query);
+            FHIRClientUtil.handleErrorResponse(response);
+            result = response.getResource(Bundle.class);
+        }
+
+        return result;
+    }
+
+    protected FHIRParameters getFHIRParameters(SearchParameterMap map) {
+        FHIRParameters result = new FHIRParameters();
+
+        Map<String, List<String>> parameters = getParameterValueMap(map);
+
+        for (Map.Entry<String, List<String>> entry : parameters.entrySet()) {
+            result.searchParam(entry.getKey(), entry.getValue().toArray(new String[entry.getValue().size()]));
+        }
+
+        if (map.count() != null) {
+            result.count(map.count());
+        }
+
+        return result;
+    }
+
+    private Map<String, List<String>> getParameterValueMap(SearchParameterMap map) {
+        Map<String, List<String>> parameters = new HashMap<>();
+        for (Map.Entry<String, List<List<? extends IQueryParameter>>> entry : map.entrySet()) {
+            for (List<? extends IQueryParameter> sublist : entry.getValue()) {
+                for (IQueryParameter p : sublist) {
+                    String name = getModifiedName(entry.getKey(), p);
+
+                    List<String> values = parameters.computeIfAbsent(name, k -> new ArrayList<>());
+                    values.add(FHIRClientUtil.urlencode(p.getParameterValue()));
+                }
+            }
+        }
+        return parameters;
+    }
+
+    protected String getModifiedName(String name, IQueryParameter param) {
```

Comment:
some comments as to the purpose of the message would be helpful.  it looks familiar, and there is an opportunity to deduplicate possibly.